be/src/storage/lake/primary_key_compaction_policy.h

```diff
@@ -75,41 +75,54 @@ class RowsetCandidate {
 };
 
 struct PKSizeTieredLevel {
-    PKSizeTieredLevel(const std::vector<RowsetCandidate>& rs, bool print_log) : rowsets(rs.begin(), rs.end()) {
-        calc_compaction_score(rs, print_log);
+    PKSizeTieredLevel(const std::vector<RowsetCandidate>& rs, int64_t compact_level)
+            : rowsets(rs.begin(), rs.end()), compact_level(compact_level) {
+        calc_compaction_score(rs);
     }
-    PKSizeTieredLevel(const PKSizeTieredLevel& level) : rowsets(level.rowsets), score(level.score) {}
+    PKSizeTieredLevel(const PKSizeTieredLevel& level)
+            : rowsets(level.rowsets), score(level.score), compact_level(level.compact_level) {}
 
     // caculate the score of this level.
-    void calc_compaction_score(const std::vector<RowsetCandidate>& rs, bool print_log) {
+    void calc_compaction_score(const std::vector<RowsetCandidate>& rs) {
         std::stringstream debug_ss;
         for (const auto& rowset : rs) {
             score += rowset.score;
             debug_ss << "[Rowset: " << rowset.rowset_meta_ptr->id() << " Size: " << rowset.rowset_meta_ptr->data_size()
                      << " Rows: " << rowset.rowset_meta_ptr->num_rows()
                      << " Dels: " << rowset.rowset_meta_ptr->num_dels() << " Score: " << rowset.score << "] ";
         }
-        if (print_log) {
-            VLOG(2) << "PKSizeTieredLevel " << debug_ss.str();
-        }
+        VLOG(2) << "PKSizeTieredLevel " << debug_ss.str();
     }
 
-    // Merge another level's top rowset
-    void merge_top(const PKSizeTieredLevel& other) {
-        DCHECK(other.rowsets.size() == 1);
-        if (!other.rowsets.empty()) {
+    // Merge another level's rowset
+    void merge_level(PKSizeTieredLevel& other) {
+        while (!other.rowsets.empty()) {
             const auto& top_rowset = other.rowsets.top();
-            // Rowset with overlap segments shouldn't be merged here.
-            DCHECK(!top_rowset.multi_segment_with_overlapped());
             rowsets.push(top_rowset);
             score += top_rowset.score;
+            other.rowsets.pop();
         }
     }
 
+    // Add other level's rowset into candidate rowsets
+    void add_candidate_rowsets(PKSizeTieredLevel& other) {
+        while (!other.rowsets.empty()) {
+            const auto& top_rowset = other.rowsets.top();
+            candidate_rowsets.push_back(top_rowset);
+            other.rowsets.pop();
+        }
+    }
+
+    void set_compact_leve(int64_t compact_level) { this->compact_level = compact_level; }
+
+    int64_t get_compact_level() { return compact_level; }
+
     bool operator<(const PKSizeTieredLevel& other) const { return score < other.score; }
 
     std::priority_queue<RowsetCandidate> rowsets;
+    std::vector<RowsetCandidate> candidate_rowsets;
```

Comment:
I think `other_level_rowsets` might be clearer?

Reply:
> I think `other_level_rowsets` might be clearer?

update.